### PR TITLE
[citrix_adc] improve grok parsing in sslvpn pipeline

### LINKED ISSUE #11312: [citrix_adc]: SSLVPN pipeline Grok pattern failure for USER
### Integration Name

Citrix ADC [citrix_adc]

### Dataset Name

citrix_adc.log

### Integration Version

1.8.1

### Agent Version

8.15.1

### Agent Output Type

elasticsearch

### Elasticsearch Version

8.15.1

### OS Version and Architecture

ECK on x86_64

### Software/API Version

_No response_

### Error Message

No message, but tracked down the problem to the `logs-citrix_adc.log-1.8.1-sslvpn_and_aaatm_feature` pipeline, the 5th pattern in the grok pipeline: 

```
^User %{USER:citrix_adc.log.user} - Client_ip %{IP:citrix_adc.log.client_ip} - Nat_ip %{IP:citrix_adc.log.nat.ip} - Vserver %{IP:citrix_adc.log.vserver.ip}:%{INT:citrix_adc.log.vserver.port} - Source %{IP:citrix_adc.log.source.ip}:%{INT:citrix_adc.log.source.port} - Destination %{IP:citrix_adc.log.destination.ip}:%{INT:citrix_adc.log.destination.port} - Start_time \"%{DATA:citrix_adc.log.start_time}\" - End_time \"%{GREEDYDATA:citrix_adc.log.end_time}\" - Duration %{DATA:citrix_adc.log.duration} - Total_bytes_send %{NUMBER:citrix_adc.log.total_bytes_send:int} - Total_bytes_recv %{NUMBER:citrix_adc.log.total_bytes_received:int} - Total_compressedbytes_send %{NUMBER:citrix_adc.log.total_compressed_bytes_send:int} - Total_compressedbytes_recv %{NUMBER:citrix_adc.log.total_compressed_bytes_recieved:int} - Compression_ratio_send %{NUMBER:citrix_adc.log.compression_ratio_send:float}% - Compression_ratio_recv %{NUMBER:citrix_adc.log.compression_ratio_recieved:float}% - Access %{WORD:citrix_adc.log.access} - Group\(s\) \"%{DATA:citrix_adc.log.groups}\"$
```

### Event Original

```
{
      "citrix": {
        "extended": {
          "message": "User redacted03@redacted.tld - Client_ip 10.0.0.1 - Nat_ip 10.0.0.1 - Vserver 172.16.0.1:443 - Source 10.0.0.1:49264 - Destination 192.168.106.5:80 - Start_time \"2024/09/30:09:23:05 \" - End_time \"2024/09/30:09:23:06 \" - Duration 00:00:01  - Total_bytes_send 1180 - Total_bytes_recv 1272 - Total_compressedbytes_send 0 - Total_compressedbytes_recv 0 - Compression_ratio_send 0.00% - Compression_ratio_recv 0.00% - Access Allowed - Group(s) \"N/A\""
        }
      }
    }
```

(Sensitive bits sanitized)

### What did you do?

I sample-tested this log against the `logs-citrix_adc.log-1.8.1-sslvpn_and_aaatm_feature` pipeline.

### What did you see?

The email-based username was not accepted by the grok pattern.

### What did you expect to see?

The email-based username should have been accepted.

### Anything else?

I tested this document and pattern in the Grok Simulator in the Dev Console.  It turns out that the very first pattern, 
`%{USER:citrix_adc.log.user}`, does not match the user in the log line: `redacted03@redacted.tld` (sanitized). 

If I changed it to `%{DATA:citrix_adc.log.user}`, it worked fine. It was then that I looked up the [grok patterns](https://github.com/logstash-plugins/logstash-patterns-core/blob/main/patterns/ecs-v1/grok-patterns):

```
USERNAME [a-zA-Z0-9._-]+
USER %{USERNAME}
EMAILLOCALPART [a-zA-Z0-9!#$%&'*+\-/=?^_`{|}~]{1,64}(?:\.[a-zA-Z0-9!#$%&'*+\-/=?^_`{|}~]{1,62}){0,63}
EMAILADDRESS %{EMAILLOCALPART}@%{HOSTNAME}
```

`USER` is an alias of `USERNAME`, and `USERNAME` will not recognize the `@` character as valid, so the grok pattern fails.

Possible solutions include just using `DATA` here, or making the grok more complicated than it needs to be to accept either `USER` or `EMAILADDRESS` as valid patterns.

### PR DESCRIPTION
**Pull request by mjwolf**
## Proposed commit message

Improve the grok processing in the sslvpn pipeline to properly parse more test cases:
- username may include domain name, ie 'username' or 'username@domain.com', this will now parse both
- Some logs did not include spaces at the end of lines or between fields, the parse will now handle these
- Add parsing of `Message` log type
- Fix mis-spelling of recieved -> received in field name
  - This could be considered a breaking change, so this is why there's a major version increase here.
- Change all `on_failure` messages to include tag and error details
- Add support for some more message variations that have been observed.

This also refactors the grok patterns to split them up by message type. This should make things more maintainable, by making it easier to find the pattern that's expected to match each type.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
- [x] I have verified that any added dashboard complies with Kibana's [Dashboard good practices](https://docs.elastic.dev/ux-guidelines/data-viz/dashboard-good-practices) 

Fixes #11312

### CODE REVIEW

File: packages/citrix_adc/data_stream/log/elasticsearch/ingest_pipeline/native.yml [hunk 1]
```diff
@@ -337,7 +337,7 @@ processors:
       field:
         - citrix_adc.log.action
         - citrix_adc.log.browser
-        - citrix_adc.log.bytes.received
+        - citrix_adc.log.bytes.recieved
```
Comment (taylor-swanson) [inline packages/citrix_adc/data_stream/log/elasticsearch/ingest_pipeline/native.yml:340]:
```suggestion
        - citrix_adc.log.bytes.received
```

This was misspelled in the fields definitions, so unfortunately, the misspelling must remain.

**Review by taylor-swanson — COMMENTED** [review]
I'll take a closer look at the groks on Monday, but this is a good direction to go. I was just thinking we should use the log type to split up the groks.

Also, could you revert the changes to the other data streams and run `elastic-package test --generate` with the `-d log` flag so it only generates for the log data stream?

File: packages/citrix_adc/data_stream/log/_dev/test/pipeline/test-citrix-waf-native.log-expected.json [hunk 2]
```diff
@@ -3240,24 +3246,13 @@
             },
             "related": {
                 "ip": [
-                    "1.128.0.0",
-                    "192.168.1.101"
-                ],
-                "user": [
-                    "ClarkKent"
```
Comment (taylor-swanson) [inline packages/citrix_adc/data_stream/log/_dev/test/pipeline/test-citrix-waf-native.log-expected.json:3247]: Looks like we aren't handling `CLISEC_EXP_EVAL` in the sslvn_and_aaatm pipeilne

**Review by taylor-swanson — COMMENTED** [review]
For the most part, the groks seem fine.

The non-log data stream expected files are still showing up in the changed files list. Not sure if the revert didn't work or if elastic-package is re-adding them? It doesn't for me, but you have to make sure that `-d log` is added to the list of flags for `elastic-package test -g -d log`

File: packages/citrix_adc/data_stream/log/elasticsearch/ingest_pipeline/sslvpn_and_aaatm_feature.yml [hunk 3]
```diff
@@ -1,29 +1,119 @@
...
+      patterns:
+        - '^(Sessionid|Session id) %{NUMBER:citrix_adc.log.session_id:int} - User %{DATA:citrix_adc.log.user} - Client_ip %{IP:citrix_adc.log.client_ip} - Nat_ip (%{IP:citrix_adc.log.nat.ip}|"%{DATA}") - Vserver_ip %{IP:citrix_adc.log.vserver.ip} - Errmsg "%{DATA:citrix_adc.log.errmsg}" ?$'
+
+  - grok:
+      tag: grok_sslvpn_message
+      if: 'ctx.citrix.name == "Message"'
+      field: citrix.extended.message
+      patterns:
+        - '^Logout handler : %{DATA}, for user <%{USERNAME|EMAILADDRESS:citrix_adc.log.username}>$'
+
```
Comment (taylor-swanson) [inline packages/citrix_adc/data_stream/log/elasticsearch/ingest_pipeline/sslvpn_and_aaatm_feature.yml:116]:
```yaml

  - grok:
      tag: grok_sslvpn_clisec_exp_eval
      if: 'ctx.citrix.name == "CLISEC_EXP_EVAL"'
      field: citrix.extended.message
      patterns:
        - '^User %{USER:citrix_adc.log.user}%{SPACE}:%{SPACE}- Client%{SPACE}IP %{IP:citrix_adc.log.client_ip} - Vserver %{IP:citrix_adc.log.vserver.ip}:%{INT:citrix_adc.log.vserver.port} - Client%{SPACE}security%{SPACE}check%{SPACE}Passed\(%{NUMBER:citrix_adc.log.client_security_check_status:int}\)%{SPACE}on%{SPACE}the%{SPACE}client%{SPACE}machine$'

```

File: packages/citrix_adc/data_stream/log/_dev/test/pipeline/test-citrix-waf-native.log-expected.json (hunk 2 above)
Comment (taylor-swanson) [reply packages/citrix_adc/data_stream/log/_dev/test/pipeline/test-citrix-waf-native.log-expected.json:3247]:
```yaml

  - grok:
      tag: grok_sslvpn_clisec_exp_eval
      if: 'ctx.citrix.name == "CLISEC_EXP_EVAL"'
      field: citrix.extended.message
      patterns:
        - '^User %{USER:citrix_adc.log.user}%{SPACE}:%{SPACE}- Client%{SPACE}IP %{IP:citrix_adc.log.client_ip} - Vserver %{IP:citrix_adc.log.vserver.ip}:%{INT:citrix_adc.log.vserver.port} - Client%{SPACE}security%{SPACE}check%{SPACE}Passed\(%{NUMBER:citrix_adc.log.client_security_check_status:int}\)%{SPACE}on%{SPACE}the%{SPACE}client%{SPACE}machine$'

```

File: packages/citrix_adc/data_stream/log/elasticsearch/ingest_pipeline/alg_feature.yml [hunk 4]
```diff
@@ -183,4 +183,5 @@ on_failure:
       value: pipeline_error
   - append:
       field: error.message
-      value: '{{{ _ingest.on_failure_message }}}'
+      value: |-
+        Processor "{{{ _ingest.on_failure_processor_type }}}" with tag "{{{ _ingest.on_failure_processor_tag }}}" in pipeline "{{{ _ingest.on_failure_pipeline }}}" failed with message "{{{ _ingest.on_failure_message }}}"
```
Comment (dwhyrock) [inline packages/citrix_adc/data_stream/log/elasticsearch/ingest_pipeline/alg_feature.yml:187]: These `error.message` values seem pretty helpful.  Are they mostly helpful if they are not in the main `default.yml` processor?

Should we mimic this behavior in integrations going forward?

File: packages/citrix_adc/data_stream/log/elasticsearch/ingest_pipeline/sslvpn_and_aaatm_feature.yml [hunk 5]
```diff
@@ -1,29 +1,119 @@
...
 processors:
+  # Keep the grok patterns separated by type, for efficiency and maintainability
   - grok:
-      description: Extract fields from message of SSLVPN and AAATM feature.
-      tag: grok_sslvpn_and_aaatm_feature
+      tag: grok_sslvpn_login
+      if: 'ctx.citrix.name == "LOGIN"'
       field: citrix.extended.message
       patterns:
-        - '^User %{USER:citrix_adc.log.user} - Client_ip %{IP:citrix_adc.log.client_ip} - Nat_ip %{IP:citrix_adc.log.nat.ip} - Vserver %{IP:citrix_adc.log.vserver.ip}:%{INT:citrix_adc.log.vserver.port} - Browser_type \"%{DATA:citrix_adc.log.browser_type}\" - SSLVPN_client_type %{DATA:citrix_adc.log.sslvpn_client_type} - Group\(s\) \"%{DATA:citrix_adc.log.groups}\"$'
```
Comment (dwhyrock) [inline packages/citrix_adc/data_stream/log/elasticsearch/ingest_pipeline/sslvpn_and_aaatm_feature.yml:9]: This all as one block definitely needed to be broken up.  Good job tackling that.

**Review by dwhyrock — COMMENTED** [review]
A couple minor comments/questions.  I didn't approve since it looks like @taylor-swanson has a request for a change.

File: packages/citrix_adc/data_stream/log/elasticsearch/ingest_pipeline/alg_feature.yml (hunk 4 above)
Comment (taylor-swanson) [reply packages/citrix_adc/data_stream/log/elasticsearch/ingest_pipeline/alg_feature.yml:187]: It's per pipeline, so yeah we have to add it to the sub-pipelines as well (it won't unwind up the default pipeline or anything like that).

**Comment by taylor-swanson** [comment]
> A couple minor comments/questions. I didn't approve since it looks like @taylor-swanson has a request for a change.

It looks like @mjwolf is out today, I'll make the necessary changes and get them pushed up.

**Review by dwhyrock — APPROVED** [review]

**Review by taylor-swanson — APPROVED** [review]
(approving Michael's changes)

**Pull request merged**